mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""Contains classes for observations and suggestions."""
+
+from __future__ import annotations
+
+from typing import Any, Iterator, List, Optional
+
+import pandas as pd
+from ConfigSpace import Configuration, ConfigurationSpace
+
+
+def compare_optional_series(left: Optional[pd.Series], right: Optional[pd.Series]) -> bool:
+    """
+    Compare Series that may also be None.
+
+    Parameters
+    ----------
+    left : Optional[pd.Series]
+        The left Series to compare
+    right : Optional[pd.Series]
+        The right Series to compare
+
+    Returns
+    -------
+    bool
+        Compare the equality of two Optional[pd.Series] objects
+    """
+    if type(left) != type(right):
+        return False
+    elif left is not None and right is not None:
+        if not left.equals(right):
+            return False
+    return True
+
+
+def compare_optional_dataframe(
+    left: Optional[pd.DataFrame], right: Optional[pd.DataFrame]
```

Comment:
```suggestion
    left: Optional[pd.DataFrame], right: Optional[pd.DataFrame],
```

nit: add a trialing comma and let black reformat to multiple lines